torch/csrc/distributed/rpc/torchscript_functions.cpp

```diff
@@ -72,25 +70,61 @@ std::shared_ptr<UserRRef> remoteTorchscript(
       returns.size());
   auto returnType = returns.at(0).type();
 
-  auto userRRefPtr = ctx.createUserRRef(dstWorkerInfo.id_, returnType);
+  if (ctx.getWorkerId() != dstWorkerInfo.id_) {
+    auto userRRefPtr = ctx.createUserRRef(dstWorkerInfo.id_, returnType);
 
-  auto scriptRemoteCall = std::make_unique<ScriptRemoteCall>(
-      qualifiedName,
-      std::move(stack),
-      userRRefPtr->rrefId(),
-      userRRefPtr->forkId());
+    auto scriptRemoteCall = std::make_unique<ScriptRemoteCall>(
+        qualifiedName,
+        std::move(stack),
+        userRRefPtr->rrefId(),
+        userRRefPtr->forkId());
 
-  auto fm = torch::distributed::autograd::sendMessageWithAutograd(
-      *rpcAgentPtr,
-      dstWorkerInfo,
-      std::move(*scriptRemoteCall).toMessage(),
-      false,
-      nullptr);
+    auto fm = torch::distributed::autograd::sendMessageWithAutograd(
+        *rpcAgentPtr,
+        dstWorkerInfo,
+        std::move(*scriptRemoteCall).toMessage(),
+        true /*forceGradRecording*/,
+        nullptr);
+
+    ctx.addPendingUser(userRRefPtr->forkId(), userRRefPtr);
+    fm->addCallback(callback::confirmPendingUser);
+
+    return userRRefPtr;
+  } else {
+    auto ownerRRefPtr = ctx.createOwnerRRef(returnType);
+    // prevent this owner RRef be deleted due to other forks
+    ctx.addSelfAsFork(ownerRRefPtr);
+
+    auto scriptRemoteCall = std::make_unique<ScriptRemoteCall>(
+        qualifiedName,
+        std::move(stack),
+        ownerRRefPtr->rrefId(),
+        ownerRRefPtr->rrefId());
 
-  ctx.addPendingUser(userRRefPtr->forkId(), userRRefPtr);
-  fm->addCallback(callback::confirmPendingUser);
+    auto fm = torch::distributed::autograd::sendMessageWithAutograd(
+        *rpcAgentPtr,
+        dstWorkerInfo,
+        std::move(*scriptRemoteCall).toMessage(),
+        true /*forceGradRecording*/,
+        nullptr);
 
-  return userRRefPtr;
+    // The callback is the same as finishCreatingOwnerRRef() in
+    // python_functions.py, the difference is that ownerRRefPtr created here
+    // always holds an IValue, no need to explicitly hold GIL and clear
+    // deletedRRef. Not calling finishCreatingOwnerRRef() because it is not
+    // necessary to introduce python dependency in this file.
+    fm->addCallback([](const Message& message,
+                       const c10::optional<utils::FutureError>& futErr) {
+      RRefContext::handleException(futErr);
+      auto rr = RemoteRet::fromMessage(message);
+      TORCH_INTERNAL_ASSERT(
+          rr->rrefId() == rr->forkId(),
+          "Expecting an OwnerRRef as RemoteRet but got a fork.");
+      auto& refCtx = RRefContext::getInstance();
+      refCtx.delForkOfOwner(rr->rrefId(), rr->rrefId());
```

Comment:
I prefer to refactor `finishCreatingOwnerRRef()` into 
`finishCreatingOwnerRRef()` (in rref_ccontetx.cpp callback namespace) and `finishCreatingOwnerRRefWithGIL()` (in python_functions) to reduce duplicate code.

Reply:
future call back schema is defined as void function(const message&, FutureError&), finishCreatingOwnerRRefWithGIL() can not call finishCreatingOwnerRRef(), as it needs finishCreatingOwnerRRef() return deletedRRef?

Reply:
Maybe this way?

std::shared_ptr<RRef> finishCreatingOwnerRRef() {

}

For Python remote call
```
    fm->addCallback([](const Message& message,
                       const c10::optional<utils::FutureError>& futErr) {
      auto deletedRRef = callback::finishCreatingOwnerRRef(message, futErr);
     if (deletedRRef && deletedRRef->isPyObj()) {
          AutoGil gil;
          deletedRRef.reset()
     }
```

For Script remote call
```
    fm->addCallback([](const Message& message,
                       const c10::optional<utils::FutureError>& futErr) {
      callback::finishCreatingOwnerRRef(message, futErr);
```

Reply:
sounds good!